File: src/engine/project.ts
// Default project factory — pure, so it's easy to test and to reset from.

import type { ProjectState, Pattern, VoiceState } from "./types";
import { neutralTimeMap } from "./timemap";
import { scrambleSteps } from "./patterncmd";

export const STEP_COUNT = 16;

/**
 * How many Voices a project has, and the range one may have.
 *
 * M shipped four and Classic still opens with four — the window is drawn for
 * four lanes and that is the interface being recreated. But four was never a
 * property of the music, only of the 1986 product, and the engine has no reason
 * to know the number: a project carries as many Voices as its `voices` array is
 * long.
 *
 * There is deliberately no `voiceCount` field. A stored count can disagree with
 * the array it counts, and then every reader has to decide which to believe.
 * `voices.length` is the count; `voiceCount()` below is how to ask.
 */
export const DEFAULT_VOICE_COUNT = 4;
export const MIN_VOICE_COUNT = 1;
export const MAX_VOICE_COUNT = 16;

/** @deprecated Prefer `voiceCount(project)`. Kept for the four-lane Classic UI. */
export const VOICE_COUNT = DEFAULT_VOICE_COUNT;

/** How many Voices this project has. The array is the authority. */
export function voiceCount(project: { voices: readonly unknown[] }): number {
  return project.voices.length;
}

export function clampVoiceCount(value: number): number {
  const rounded = Math.round(value);
  if (!Number.isFinite(rounded)) return DEFAULT_VOICE_COUNT;
  return Math.max(MIN_VOICE_COUNT, Math.min(MAX_VOICE_COUNT, rounded));
}
export const PATTERN_COUNT = 24;
export const CYCLIC_NEUTRAL_LEVEL = 2;
/** Default ceiling of the Pattern Size Numerical, as M shipped it. */
export const DEFAULT_MAX_SIZE = 100;

/** A C-major seed riff so the app makes sound the moment you press play. */
const SEED_RIFF = [60, 62, 64, 65, 67, 69, 71, 72];

export function createDefaultPattern(id: string, seedRiff = false): Pattern {
  const steps = Array.from({ length: STEP_COUNT }, (_, i) => ({
    pitches: seedRiff && i < SEED_RIFF.length ? [SEED_RIFF[i]] : [],
  }));
  const outputLength = seedRiff ? SEED_RIFF.length : STEP_COUNT;
  return {
    id,
    steps,
    scrambledSteps: [
      ...scrambleSteps(steps.slice(0, outputLength), 1),
      ...steps.slice(outputLength).map((step) => ({ pitches: [...step.pitches] })),
    ],
    scrambleGeneration: 0,
    outputLength,
    maxSize: DEFAULT_MAX_SIZE,
    chordMode: "single",
    insertMode: "insert",
    drumMachine: false,
    timeBaseNumerator: 1,
    timeBaseDenominator: 8,
    phase: 0,
  };
}

export function createDefaultVoice(index: number): VoiceState {
  return {
    patternIndex: index,
    playEnabled: index === 0,
    transposition: 0,
    noteOrderMix: { original: 100, cyclic: 0, utterly: 0 },
    density: 1,
    velocityRange: { low: 48, high: 110 },
    timeBaseNumerator: 1,
    timeBaseDenominator: 8,
    phase: 0,
    timeDistort: neutralTimeMap(),
    legato: 1,
    channel: index + 1,
    outputChannels: [index + 1],
    program: 0,
    sourceChannel: "all",
    inputUse: "disabled",
    echoInput: false,
    mouseAdvance: false,
  };
}

export function createDefaultProject(voices = DEFAULT_VOICE_COUNT): ProjectState {
  const count = clampVoiceCount(voices);
  const neutralCycle = () =>
    Array.from({ length: count }, () =>
      Array(STEP_COUNT).fill(CYCLIC_NEUTRAL_LEVEL),
    );
  return {
    tempo: 120,
    patterns: Array.from({ length: PATTERN_COUNT }, (_, i) =>
      createDefaultPattern(`pattern-${i + 1}`, i === 0),
    ),
    voices: Array.from({ length: count }, (_, i) => createDefaultVoice(i)),
    root: 0,
    scale: "major",
    scaleSnap: false,
    seed: 1,
    diatonicTranspose: false,
    secondOrderTranspose: false,
    chordTones: false,
    midiAssignments: {
      inputs: Array.from({ length: 16 }, (_, i) => ({ deviceId: null, channel: i + 1 })),
      outputs: Array.from({ length: 16 }, (_, i) => ({ deviceId: null, channel: i + 1 })),
      programBase: 0,
      latencyMs: 0,
      conductXController: 16,
      conductYController: 17,
    },
    echoMapChannels: [],
    cyclic: {
      accent: neutralCycle(),
      legato: neutralCycle(),
      rhythm: neutralCycle(),
    },
    cyclicLengths: {
      accent: Array(count).fill(STEP_COUNT),
      legato: Array(count).fill(STEP_COUNT),
      rhythm: Array(count).fill(STEP_COUNT),
    },
    cyclicValues: {
      legato: [6, 25, 50, 75, 100],
      rhythm: [1, 1, 1.5, 2, 5],
    },
  };
}

File: plugin/engine/Project.cpp
#include "Project.h"

#include <string>

namespace idm
{

namespace
{

/** A C-major seed riff, so the instrument makes sound the moment you press
    play. Ported from `createDefaultPattern`. */
constexpr int seedRiff[] = { 60, 62, 64, 65, 67, 69, 71, 72 };
constexpr int seedRiffLength = 8;

/** M's stored Cyclic Random copy, built the way `scrambleSteps` builds it:
    a Fisher-Yates shuffle of the played region, seeded at 1, with the unplayed
    tail carried over untouched. */
std::vector<StepEvent> scrambleSteps (const std::vector<StepEvent>& steps, int outputLength)
{
    std::vector<StepEvent> played (steps.begin(), steps.begin() + outputLength);
    const auto order = makeCyclicOrder (outputLength, 1);

    std::vector<StepEvent> out;
    out.reserve (steps.size());

    for (const auto index : order)
        out.push_back (played[(size_t) index]);

    for (size_t i = (size_t) outputLength; i < steps.size(); ++i)
        out.push_back (steps[i]);

    return out;
}

Pattern createDefaultPattern (bool withSeedRiff)
{
    Pattern pattern;
    pattern.steps.resize (stepCount);

    if (withSeedRiff)
        for (int i = 0; i < seedRiffLength; ++i)
            pattern.steps[(size_t) i].pitches = { seedRiff[i] };

    pattern.outputLength = withSeedRiff ? seedRiffLength : stepCount;
    pattern.scrambledSteps = scrambleSteps (pattern.steps, pattern.outputLength);

    return pattern;
}

VoiceState createDefaultVoice (int index)
{
    VoiceState voice;
    voice.patternIndex = (size_t) index;
    voice.playEnabled = index == 0;
    voice.transposition = 0;
    voice.noteOrderMix = { 100, 0, 0 };
    voice.density = 1.0;
    voice.velocityRange = { 48.0, 110.0 };
    voice.timeBaseNumerator = 1.0;
    voice.timeBaseDenominator = 8.0;
    voice.phase = 0.0;
    voice.timeDistort = {};
    voice.legato = 1.0;
    voice.outputChannels = { index + 1 };
    voice.mouseAdvance = false;

    return voice;
}

} // namespace

ProjectState createDefaultProject (int voices)
{
    const auto count = voices < 1 ? 1 : (voices > 16 ? 16 : voices);

    ProjectState state;
    state.tempo = 120.0;
    state.root = 0;
    state.scale = Scale::major;
    state.scaleSnap = false;
    state.seed = 1;
    state.diatonicTranspose = false;
    state.secondOrderTranspose = false;
    state.chordTones = false;

    state.patterns.reserve (patternCount);

    for (int i = 0; i < patternCount; ++i)
        state.patterns.push_back (createDefaultPattern (i == 0));

    state.voices.reserve ((size_t) count);

    for (int i = 0; i < count; ++i)
        state.voices.push_back (createDefaultVoice (i));

    const auto neutralCycle = [count]
    {
        std::vector<std::vector<CyclicStep>> rows;
        rows.reserve ((size_t) count);

        for (int i = 0; i < count; ++i)
            rows.push_back (std::vector<CyclicStep> ((size_t) stepCount,
                                                     CyclicStep::level (cyclicNeutralLevel)));

        return rows;
    };

    state.cyclic = { neutralCycle(), neutralCycle(), neutralCycle() };

    const std::vector<int> lengths ((size_t) count, stepCount);
    state.cyclicLengths = { lengths, lengths, lengths };

    state.cyclicValues = { { 6.0, 25.0, 50.0, 75.0, 100.0 }, { 1.0, 1.0, 1.5, 2.0, 5.0 } };

    return state;
}

} // namespace idm
